File: ruby/leetcode/shortest-completing-word.rb
# @param {String} license_plate
# @param {String[]} words
# @return {String}
def shortest_completing_word(license_plate, words)
    lp = pull_letters(license_plate)
    min_length = Float::INFINITY
    shortest_word = nil
    for i in 0...words.size()
        if has_match(lp, words[i])
            if words[i].size() < min_length
                min_length = words[i].size()
                shortest_word = words[i]
            end
        end
    end
    return shortest_word
end

def pull_letters(license_plate)
    license_plate.downcase.gsub(/[\d|\s]?/, '').split('')
end

def has_match(lp, word)
    candidates = lp.clone()
    for i in 0...word.size()
        char_index = candidates.index(word[i])
        if char_index
            candidates.delete_at(char_index)
        end
        return true if candidates.size() === 0
    end
    return false
end

File: javascript/leetcode/shortest-completing-word.js
/**
 * @param {string} licensePlate
 * @param {string[]} words
 * @return {string}
 */
var shortestCompletingWord = function(licensePlate, words) {
    const lp = pullLetters(licensePlate);
    let minLen = Infinity;
    let shortestWord;
    for (var i = 0; i < words.length; i++) {
        if (hasMatch(lp, words[i])) {
            if (words[i].length < minLen) {
                minLen = words[i].length;
                shortestWord = words[i]; 
            }
        }
    }
    return shortestWord;
};

var pullLetters = function(licensePlate) {
    // downcases letters and strips numbers
    return licensePlate
      .toLowerCase()
      .replace(/[\d|\s]?/g, '')
      .split('');
}

var hasMatch = function(lp, word) {
    // checks if all letters contained in lp char array
    let candidates = lp.slice(0)
    for (var i = 0; i < word.length; i++) {
        let charIndex = candidates.indexOf(word[i]);
        if (charIndex !== -1) {
            candidates.splice(charIndex, 1);
        }
        if (candidates.length === 0) { return true; }
    }
    return false;
}
